### PR TITLE
On Solana, stack allocations must happen in the first block

### PR DESCRIPTION
**Pull request by seanyoung**
If not, the Solana llvm target will fail to generate code with the error:

Unsupported dynamic stack allocation on create_program_address

Fixes https://github.com/hyperledger/solang/issues/1019

Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

File: src/emit/instructions.rs [hunk 1]
```diff
@@ -595,7 +595,7 @@ pub(super) fn process_instruction<'a, T: TargetRuntime<'a> + ?Sized>(
                 }
             }
 
-            let ret = target.builtin_function(bin, func, &parms, ns);
+            let ret = target.builtin_function(bin, function, func, &parms, ns);
```
Comment (LucasSte) [inline src/emit/instructions.rs:598]: I believe the bad naming here might have led to a confusion. What you think of giving better names to `func` and `function`?

File: tests/solana_tests/builtin.rs [hunk 2]
```diff
@@ -101,17 +105,21 @@ fn pda() {
                 return create_program_address([a, b], program_id);
             }
 
-            function create_pda2_bump() public returns (address, bytes1) {
+            function create_pda2_bump(bool cond) public returns (address, bytes1) {
```
Comment (LucasSte) [inline tests/solana_tests/builtin.rs:108]: What do you want to achieve with this boolean here?
All your tests set this argument as `true`.

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: tests/solana_tests/builtin.rs (hunk 2 above)
Comment (seanyoung) [reply tests/solana_tests/builtin.rs:108]: This is to ensure that `create_program_address()` is not in the first block of the cfg; that is the problem case.

File: src/emit/instructions.rs (hunk 1 above)
Comment (seanyoung) [reply src/emit/instructions.rs]: I've renamed them

**Review by xermicus — APPROVED** [review]

File: tests/solana_tests/builtin.rs (hunk 2 above)
Comment (LucasSte) [reply tests/solana_tests/builtin.rs:108]: I think we should still be testing `create_program_address` as the first instruction in a block.
What about this?

```solidity
function create_pda(bool cond) public returns (address) {
                 address program_id = address"BPFLoaderUpgradeab1e11111111111111111111111";
                 address addr =  create_program_address(["Talking", "Cats"], program_id);
                 if (cond) {
                     return create_program_address(["Talking", "Squirrels"], program_id);
                 } else {
                     return addr;
                 }
             }
```

Comment (seanyoung) [reply tests/solana_tests/builtin.rs:108]: I've added this test

**Review by LucasSte — APPROVED** [review]

**Pull request merged**